File: composeApp/src/androidMain/kotlin/com/kissangram/viewmodel/EditProfileViewModel.kt
package com.kissangram.viewmodel

import android.app.Application
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.kissangram.model.User
import com.kissangram.model.UserRole
import com.kissangram.repository.AndroidAuthRepository
import com.kissangram.repository.AndroidCropsRepository
import com.kissangram.repository.AndroidLocationRepository
import com.kissangram.repository.AndroidPreferencesRepository
import com.kissangram.repository.FirestoreUserRepository
import com.kissangram.usecase.GetAllCropsUseCase
import com.kissangram.usecase.GetCurrentUserUseCase
import com.kissangram.usecase.GetDistrictsUseCase
import com.kissangram.usecase.GetStatesUseCase
import com.kissangram.usecase.UpdateProfileUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class EditProfileViewModel(
    application: Application
) : AndroidViewModel(application) {
    
    // Repositories
    private val prefs = AndroidPreferencesRepository(application.applicationContext)
    private val authRepository = AndroidAuthRepository(
        context = application.applicationContext,
        activity = null,
        preferencesRepository = prefs
    )
    private val userRepository = FirestoreUserRepository(authRepository = authRepository)
    private val locationRepository = AndroidLocationRepository()
    private val cropsRepository = AndroidCropsRepository()
    
    // Use cases
    private val getCurrentUserUseCase = GetCurrentUserUseCase(userRepository)
    private val updateProfileUseCase = UpdateProfileUseCase(userRepository)
    private val getStatesUseCase = GetStatesUseCase(locationRepository)
    private val getDistrictsUseCase = GetDistrictsUseCase(locationRepository)
    private val getAllCropsUseCase = GetAllCropsUseCase(cropsRepository)
    
    // UI State
    private val _uiState = MutableStateFlow(EditProfileUiState())
    val uiState: StateFlow<EditProfileUiState> = _uiState.asStateFlow()
    
    init {
        loadCurrentUser()
        loadStates()
        loadCrops()
    }
    
    /**
     * Load current user profile from Firestore
     */
    fun loadCurrentUser() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoadingUser = true, userError = null)
            
            try {
                val user = getCurrentUserUseCase()
                println("✅ EditProfileViewModel: Loaded user = ${user?.name}")
                
                if (user != null) {
                    _uiState.value = _uiState.value.copy(
                        currentUser = user,
                        name = user.name,
                        bio = user.bio ?: "",
                        profileImageUrl = user.profileImageUrl,
                        selectedRole = user.role,
                        selectedState = user.location?.state,
                        selectedDistrict = user.location?.district,
                        village = user.location?.village ?: "",
                        selectedCrops = user.expertise.toSet(),
                        isLoadingUser = false
                    )
                    
                    // Load districts if state is already selected
                    user.location?.state?.let { state ->
                        loadDistricts(state)
                    }
                } else {
                    _uiState.value = _uiState.value.copy(
                        isLoadingUser = false,
                        userError = "User not found"
                    )
                }
            } catch (e: Exception) {
                println("❌ EditProfileViewModel: Error loading user - ${e.message}")
                e.printStackTrace()
                _uiState.value = _uiState.value.copy(
                    isLoadingUser = false,
                    userError = e.message ?: "Failed to load user"
                )
            }
        }
    }
    
    /**
     * Load all available states from Firestore
     */
    fun loadStates() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoadingStates = true, statesError = null)
            
            try {
                val states = getStatesUseCase()
                println("✅ EditProfileViewModel: Loaded ${states.size} states")
                _uiState.value = _uiState.value.copy(
                    states = states,
                    isLoadingStates = false
                )
            } catch (e: Exception) {
                println("❌ EditProfileViewModel: Error loading states - ${e.message}")
                e.printStackTrace()
                _uiState.value = _uiState.value.copy(
                    isLoadingStates = false,
                    statesError = e.message ?: "Failed to load states"
                )
            }
        }
    }
    
    /**
     * Load districts for the selected state
     */
    fun loadDistricts(state: String) {
        if (state.isBlank()) {
            _uiState.value = _uiState.value.copy(districts = emptyList())
            return
        }
        
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoadingDistricts = true, districtsError = null)
            
            try {
                val districts = getDistrictsUseCase(state)
                _uiState.value = _uiState.value.copy(
                    districts = districts,
                    isLoadingDistricts = false
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoadingDistricts = false,
                    districtsError = e.message ?: "Failed to load districts"
                )
            }
        }
    }
    
    /**
     * Called when user selects a state
     */
    fun onStateSelected(state: String) {
        _uiState.value = _uiState.value.copy(
            selectedState = state,
            selectedDistrict = null, // Reset district when state changes
            districts = emptyList()
        )
        loadDistricts(state)
    }
    
    /**
     * Called when user selects a district
     */
    fun onDistrictSelected(district: String) {
        _uiState.value = _uiState.value.copy(selectedDistrict = district)
    }
    
    /**
     * Clear location selection
     */
    fun clearLocationSelection() {
        _uiState.value = _uiState.value.copy(
            selectedState = null,
            selectedDistrict = null,
            districts = emptyList()
        )
    }
    
    /**
     * Load all available crops from Firestore
     */
    fun loadCrops() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoadingCrops = true, cropsError = null)
            
            try {
                val crops = getAllCropsUseCase()
                println("✅ EditProfileViewModel: Loaded ${crops.size} crops")
                _uiState.value = _uiState.value.copy(
                    allCrops = crops,
                    isLoadingCrops = false
                )
            } catch (e: Exception) {
                println("❌ EditProfileViewModel: Error loading crops - ${e.message}")
                e.printStackTrace()
                _uiState.value = _uiState.value.copy(
                    isLoadingCrops = false,
                    cropsError = e.message ?: "Failed to load crops"
                )
            }
        }
    }
    
    // ==================== Profile Field Updates ====================
    
    /**
     * Update name field
     */
    fun updateName(name: String) {
        _uiState.value = _uiState.value.copy(name = name)
    }
    
    /**
     * Update bio field
     */
    fun updateBio(bio: String) {
        _uiState.value = _uiState.value.copy(bio = bio)
    }
    
    /**
     * Update profile image URL (after image selection)
     */
    fun updateProfileImageUri(uri: Uri?) {
        _uiState.value = _uiState.value.copy(selectedImageUri = uri)
    }
    
    /**
     * Update village field
     */
    fun updateVillage(village: String) {
        _uiState.value = _uiState.value.copy(village = village)
    }
    
    /**
     * Called when user selects a role
     */
    fun onRoleSelected(role: UserRole) {
        _uiState.value = _uiState.value.copy(selectedRole = role)
    }
    
    /**
     * Toggle crop selection
     */
    fun toggleCrop(crop: String) {
        val currentCrops = _uiState.value.selectedCrops
        val newCrops = if (currentCrops.contains(crop)) {
            currentCrops - crop
        } else {
            currentCrops + crop
        }
        _uiState.value = _uiState.value.copy(selectedCrops = newCrops)
    }
    
    // ==================== Save Profile ====================
    
    /**
     * Save all profile changes to Firestore
     */
    fun saveProfile(
        onSuccess: () -> Unit,
        onError: (String) -> Unit
    ) {
        viewModelScope.launch {
            val state = _uiState.value
            
            // Validate
            if (state.name.isBlank()) {
                onError("Name cannot be empty")
                return@launch
            }
            
            _uiState.value = state.copy(isSaving = true, saveError = null)
            
            try {
                println("📝 EditProfileViewModel: Saving profile...")
                
                // TODO: Upload image to Firebase Storage if selectedImageUri is set
                // For now, we'll skip image upload and use existing URL
                val imageUrl = state.profileImageUrl
                
                updateProfileUseCase(
                    name = state.name,
                    bio = state.bio.ifBlank { null },
                    profileImageUrl = imageUrl,
                    role = state.selectedRole,
                    state = state.selectedState,
                    district = state.selectedDistrict,
                    village = state.village.ifBlank { null },
                    crops = state.selectedCrops.toList().ifEmpty { null }
                )
                
                println("✅ EditProfileViewModel: Profile saved successfully!")
                _uiState.value = _uiState.value.copy(isSaving = false, saveSuccess = true)
                onSuccess()
                
            } catch (e: IllegalArgumentException) {
                println("❌ EditProfileViewModel: Validation error - ${e.message}")
                _uiState.value = _uiState.value.copy(
                    isSaving = false,
                    saveError = e.message ?: "Validation failed"
                )
                onError(e.message ?: "Validation failed")
                
            } catch (e: Exception) {
                println("❌ EditProfileViewModel: Save error - ${e.message}")
                e.printStackTrace()
                _uiState.value = _uiState.value.copy(
                    isSaving = false,
                    saveError = e.message ?: "Failed to save profile"
                )
                onError(e.message ?: "Failed to save profile")
            }
        }
    }
    
    /**
     * Clear save success flag
     */
    fun clearSaveSuccess() {
        _uiState.value = _uiState.value.copy(saveSuccess = false)
    }
    
    /**
     * Check if there are unsaved changes
     */
    fun hasUnsavedChanges(): Boolean {
        val state = _uiState.value
        val user = state.currentUser ?: return false
        
        // Check if any field has changed
        val nameChanged = state.name != user.name
        val bioChanged = state.bio != (user.bio ?: "")
        val roleChanged = state.selectedRole != user.role
        val stateChanged = state.selectedState != user.location?.state
        val districtChanged = state.selectedDistrict != user.location?.district
        val villageChanged = state.village != (user.location?.village ?: "")
        val cropsChanged = state.selectedCrops != user.expertise.toSet()
        val imageChanged = state.selectedImageUri != null
        
        return nameChanged || bioChanged || roleChanged || stateChanged || 
               districtChanged || villageChanged || cropsChanged || imageChanged
    }
}

data class EditProfileUiState(
    // Current user
    val currentUser: User? = null,
    
    // Editable fields
    val name: String = "",
    val bio: String = "",
    val profileImageUrl: String? = null,
    val selectedImageUri: Uri? = null,
    val village: String = "",
    val selectedRole: UserRole = UserRole.FARMER,
    val selectedCrops: Set<String> = emptySet(),
    
    // Location data
    val states: List<String> = emptyList(),
    val districts: List<String> = emptyList(),
    val selectedState: String? = null,
    val selectedDistrict: String? = null,
    
    // Crops data
    val allCrops: List<String> = emptyList(),
    
    // Loading states
    val isLoadingUser: Boolean = false,
    val isLoadingStates: Boolean = false,
    val isLoadingDistricts: Boolean = false,
    val isLoadingCrops: Boolean = false,
    val isSaving: Boolean = false,
    
    // Errors
    val userError: String? = null,
    val statesError: String? = null,
    val districtsError: String? = null,
    val cropsError: String? = null,
    val saveError: String? = null,
    
    // Success state
    val saveSuccess: Boolean = false
)

File: iosApp/iosApp/viewmodel/EditProfileViewModel.swift
import Foundation
import SwiftUI
import os.log
import Shared
import PhotosUI

private let editProfileVMLog = Logger(subsystem: "com.kissangram", category: "EditProfileViewModel")

@MainActor
class EditProfileViewModel: ObservableObject {
    
    // Repositories
    private let preferencesRepository = IOSPreferencesRepository()
    private lazy var authRepository: IOSAuthRepository = {
        IOSAuthRepository(preferencesRepository: preferencesRepository)
    }()
    private lazy var userRepository: FirestoreUserRepository = {
        FirestoreUserRepository(authRepository: authRepository)
    }()
    private let locationRepository = IOSLocationRepository()
    private let cropsRepository = IOSCropsRepository()
    private let storageRepository = IOSStorageRepository()
    
    // Current user
    @Published var currentUser: User? = nil
    
    // Editable fields
    @Published var name: String = ""
    @Published var username: String = ""
    @Published var bio: String = ""
    @Published var profileImageUrl: String? = nil
    @Published var selectedImageItem: PhotosPickerItem? = nil
    @Published var selectedImageData: Data? = nil
    @Published var village: String = ""
    @Published var selectedRole: UserRole = .farmer
    @Published var selectedCrops: Set<String> = []
    
    // Location state
    @Published var states: [String] = []
    @Published var districts: [String] = []
    @Published var selectedState: String? = nil
    @Published var selectedDistrict: String? = nil
    
    // Crops state
    @Published var allCrops: [String] = []
    
    // Loading states
    @Published var isLoadingUser: Bool = false
    @Published var isLoadingStates: Bool = false
    @Published var isLoadingDistricts: Bool = false
    @Published var isLoadingCrops: Bool = false
    @Published var isLoadingLocation: Bool = false
    @Published var isSaving: Bool = false
    @Published var isUploadingImage: Bool = false
    
    // Errors
    @Published var userError: String? = nil
    @Published var statesError: String? = nil
    @Published var districtsError: String? = nil
    @Published var cropsError: String? = nil
    @Published var locationError: String? = nil
    @Published var saveError: String? = nil
    
    // Success state
    @Published var saveSuccess: Bool = false
    
    init() {
        loadCurrentUser()
        loadStates()
        loadCrops()
    }
    
    /// Load current user profile from Firestore
    func loadCurrentUser() {
        isLoadingUser = true
        userError = nil
        
        Task {
            do {
                let user = try await userRepository.getCurrentUser()
                
                await MainActor.run {
                    if let user = user {
                        self.currentUser = user
                        self.name = user.name
                        self.username = user.username
                        self.bio = user.bio ?? ""
                        self.profileImageUrl = user.profileImageUrl
                        self.selectedRole = user.role
                        self.selectedState = user.location?.state
                        self.selectedDistrict = user.location?.district
                        self.village = user.location?.village ?? ""
                        self.selectedCrops = Set(user.expertise)
                        
                        // Load districts if state is already selected
                        if let state = user.location?.state {
                            self.loadDistricts(for: state)
                        }
                    } else {
                        self.userError = "User not found"
                    }
                    
                    self.isLoadingUser = false
                }
            } catch {
                await MainActor.run {
                    self.isLoadingUser = false
                    self.userError = error.localizedDescription
                }
            }
        }
    }
    
    /// Load all available states from Firestore
    func loadStates() {
        isLoadingStates = true
        statesError = nil
        
        Task {
            do {
                let fetchedStates = try await locationRepository.getStates()
                await MainActor.run {
                    self.states = fetchedStates
                    self.isLoadingStates = false
                }
            } catch {
                await MainActor.run {
                    self.isLoadingStates = false
                    self.statesError = error.localizedDescription
                }
            }
        }
    }
    
    /// Load districts for the selected state
    func loadDistricts(for state: String) {
        guard !state.isEmpty else {
            districts = []
            return
        }
        
        isLoadingDistricts = true
        districtsError = nil
        
        Task {
            do {
                let fetchedDistricts = try await locationRepository.getDistricts(state: state)
                await MainActor.run {
                    self.districts = fetchedDistricts
                    self.isLoadingDistricts = false
                }
            } catch {
                await MainActor.run {
                    self.isLoadingDistricts = false
                    self.districtsError = error.localizedDescription
                }
            }
        }
    }
    
    /// Called when user selects a state
    func onStateSelected(_ state: String) {
        selectedState = state
        selectedDistrict = nil // Reset district when state changes
        districts = []
        loadDistricts(for: state)
    }
    
    /// Called when user selects a district
    func onDistrictSelected(_ district: String) {
        selectedDistrict = district
    }
    
    /// Clear location selection
    func clearLocationSelection() {
        selectedState = nil
        selectedDistrict = nil
        districts = []
    }
    
    /// Use current GPS location to automatically set state and district
    func useCurrentLocation() {
        guard !isLoadingLocation else { return }
        
        isLoadingLocation = true
        locationError = nil
        
        Task {
            // Check location permission
            if !locationRepository.hasLocationPermission() {
                do {
                    let granted = try await locationRepository.requestLocationPermission()
                    if !granted.boolValue {
                        await MainActor.run {
                            self.isLoadingLocation = false
                            self.locationError = "Location permission is required to use current location"
                        }
                        return
                    }
                } catch {
                    await MainActor.run {
                        self.isLoadingLocation = false
                        self.locationError = "Failed to request location permission: \(error.localizedDescription)"
                    }
                    return
                }
            }
            
            do {
                // Get GPS coordinates
                guard let coordinates = try await locationRepository.getCurrentLocation() else {
                    await MainActor.run {
                        self.isLoadingLocation = false
                        self.locationError = "Unable to get current location. Please try again."
                    }
                    return
                }
                
                // Reverse geocode to get location name
                let locationName = try await locationRepository.reverseGeocode(
                    latitude: coordinates.latitude,
                    longitude: coordinates.longitude
                )
                
                guard let locationName = locationName else {
                    await MainActor.run {
                        self.isLoadingLocation = false
                        self.locationError = "Unable to get location name. Please try manual selection."
                    }
                    return
                }
                
                // Parse location name to extract state and district
                // Format is typically "District, State" or "State" or "District"
                let parts = locationName.components(separatedBy: ",").map { $0.trimmingCharacters(in: .whitespaces) }
                
                // Just set the location parts directly without matching to lists
                // If format is "District, State", set district and state
                // If only one part, assume it's the state
                if parts.count >= 2 {
                    // Format: "District, State"
                    let districtName = parts[0]
                    let stateName = parts[1]
                    
                    await MainActor.run {
                        self.selectedState = stateName
                        self.selectedDistrict = districtName
                        self.districts = [] // Clear districts, will be loaded when state is set
                    }
                    
                    // Load districts for the state (so dropdown works)
                    do {
                        let fetchedDistricts = try await locationRepository.getDistricts(state: stateName)
                        await MainActor.run {
                            self.districts = fetchedDistricts
                            self.isLoadingLocation = false
                            self.locationError = nil
                        }
                    } catch {
                        await MainActor.run {
                            self.isLoadingLocation = false
                            self.locationError = nil // Still set the values even if districts load fails
                        }
                    }
                } else if parts.count == 1 {
                    // Only one part, assume it's the state
                    await MainActor.run {
                        self.selectedState = parts[0]
                        self.selectedDistrict = nil
                        self.districts = []
                        self.isLoadingLocation = false
                        self.locationError = nil
                    }
                    
                    // Try to load districts for the state
                    do {
                        let fetchedDistricts = try await locationRepository.getDistricts(state: parts[0])
                        await MainActor.run {
                            self.districts = fetchedDistricts
                        }
                    } catch {
                        // Ignore error, just set the state
                    }
                } else {
                    await MainActor.run {
                        self.isLoadingLocation = false
                        self.locationError = "Invalid location format"
                    }
                }
            } catch {
                await MainActor.run {
                    self.isLoadingLocation = false
                    self.locationError = "Failed to get location: \(error.localizedDescription)"
                }
            }
        }
    }
    
    /// Load all available crops from Firestore
    func loadCrops() {
        isLoadingCrops = true
        cropsError = nil
        
        Task {
            do {
                let fetchedCrops = try await cropsRepository.getAllCrops()
                await MainActor.run {
                    self.allCrops = fetchedCrops
                    self.isLoadingCrops = false
                }
            } catch {
                await MainActor.run {
                    self.isLoadingCrops = false
                    self.cropsError = error.localizedDescription
                }
            }
        }
    }
    
    // MARK: - Profile Field Updates
    
    /// Update name field
    func updateName(_ name: String) {
        self.name = name
    }
    
    /// Update bio field
    func updateBio(_ bio: String) {
        self.bio = bio
    }
    
    /// Update village field
    func updateVillage(_ village: String) {
        self.village = village
    }
    
    /// Set selected image data (from PhotosPicker or camera). Must be called when user picks an image so Save can upload it.
    func setSelectedImageData(_ data: Data?, item: PhotosPickerItem?) {
        selectedImageData = data
        selectedImageItem = item
    }
    
    /// Called when user selects a role
    func onRoleSelected(_ role: UserRole) {
        selectedRole = role
    }
    
    /// Toggle crop selection
    func toggleCrop(_ crop: String) {
        if selectedCrops.contains(crop) {
            selectedCrops.remove(crop)
        } else {
            selectedCrops.insert(crop)
        }
    }
    
    // MARK: - Save Profile
    
    /// Save all profile changes to Firestore
    func saveProfile(onSuccess: @escaping () -> Void, onError: @escaping (String) -> Void) {
        editProfileVMLog.info("saveProfile called")
        // Validate
        guard !name.trimmingCharacters(in: .whitespaces).isEmpty else {
            onError("Name cannot be empty")
            return
        }
        
        guard name.count >= 2 else {
            onError("Name must be at least 2 characters")
            return
        }
        
        guard bio.count <= 150 else {
            onError("Bio cannot exceed 150 characters")
            return
        }
        
        isSaving = true
        isUploadingImage = selectedImageData != nil
        saveError = nil
        
        Task {
            do {
                // Upload image to Firebase Storage if a new image was selected
                var imageUrl = profileImageUrl
                editProfileVMLog.info("Initial profileImageUrl for Firestore: \(String(describing: imageUrl))")
                
                if let imageData = selectedImageData {
                    isUploadingImage = true
                    editProfileVMLog.info("New image selected, uploading to Storage (size=\(imageData.count) bytes)")
                    
                    // Get current user ID for upload
                    do {
                        if let userId = try await authRepository.getCurrentUserId() {
                            do {
                                // Convert Data to KotlinByteArray
                                let kotlinByteArray = KotlinByteArray(size: Int32(imageData.count))
                                imageData.withUnsafeBytes { bytes in
                                    for (index, byte) in bytes.enumerated() {
                                        kotlinByteArray.set(index: Int32(index), value: Int8(bitPattern: byte))
                                    }
                                }
                                
                                imageUrl = try await storageRepository.uploadProfileImage(userId: userId, imageData: kotlinByteArray)
                                editProfileVMLog.info("Storage upload succeeded, imageUrl=\(String(describing: imageUrl))")
                            } catch {
                                editProfileVMLog.error("Storage upload failed: \(error.localizedDescription); keeping existing URL")
                                // Continue with existing URL if upload fails
                            }
                        } else {
                            editProfileVMLog.warning("getCurrentUserId returned nil, cannot upload image")
                        }
                    } catch {
                        editProfileVMLog.error("getCurrentUserId failed: \(error.localizedDescription)")
                        // Continue with existing URL if getting user ID fails
                    }
                    
                    isUploadingImage = false
                }
                
                editProfileVMLog.info("Sending to Firestore: profileImageUrl=\(String(describing: imageUrl))")
                // Save profile to Firestore
                await saveProfileToFirestore(imageUrl: imageUrl, onSuccess: onSuccess, onError: onError)
                
            } catch {
                await MainActor.run {
                    isSaving = false
                    isUploadingImage = false
                    saveError = error.localizedDescription
                    onError(error.localizedDescription)
                }
            }
        }
    }
    
    /// Helper to save profile to Firestore
    private func saveProfileToFirestore(imageUrl: String?, onSuccess: @escaping () -> Void, onError: @escaping (String) -> Void) async {
        editProfileVMLog.info("saveProfileToFirestore called with profileImageUrl=\(String(describing: imageUrl))")
        do {
            try await userRepository.updateFullProfile(
                name: name.trimmingCharacters(in: .whitespaces),
                bio: bio.isEmpty ? nil : bio,
                profileImageUrl: imageUrl,
                role: selectedRole,
                state: selectedState,
                district: selectedDistrict,
                village: village.isEmpty ? nil : village,
                crops: selectedCrops.isEmpty ? nil : Array(selectedCrops)
            )
            
            await MainActor.run {
                editProfileVMLog.info("Firestore update succeeded; profileImageUrl saved=\(String(describing: imageUrl))")
                self.isSaving = false
                self.isUploadingImage = false
                self.profileImageUrl = imageUrl // Update with new URL
                self.selectedImageData = nil // Clear selected image after upload
                self.selectedImageItem = nil
                self.saveSuccess = true
                onSuccess()
            }
        } catch {
            await MainActor.run {
                editProfileVMLog.error("Firestore update failed: \(error.localizedDescription)")
                self.isSaving = false
                self.isUploadingImage = false
                self.saveError = error.localizedDescription
                onError(error.localizedDescription)
            }
        }
    }
    
    /// Clear save success flag
    func clearSaveSuccess() {
        saveSuccess = false
    }
    
    /// Check if there are unsaved changes
    func hasUnsavedChanges() -> Bool {
        guard let user = currentUser else { return false }
        
        // Check if any field has changed
        let nameChanged = name != user.name
        let bioChanged = bio != (user.bio ?? "")
        let roleChanged = selectedRole != user.role
        let stateChanged = selectedState != user.location?.state
        let districtChanged = selectedDistrict != user.location?.district
        let villageChanged = village != (user.location?.village ?? "")
        let cropsChanged = selectedCrops != Set(user.expertise)
        let imageChanged = selectedImageData != nil || selectedImageItem != nil
        
        return nameChanged || bioChanged || roleChanged || stateChanged ||
               districtChanged || villageChanged || cropsChanged || imageChanged
    }
}
